Fix explosion dodging. Tweaks door resistance

Cyberboss opened the pull request:
Fixes #20820

:cl: Cyberboss
fix: Explosions can no longer be dodged
tweak: Airlocks are now destroyed by the same level explosion that destroys walls
tweak: Diamond/External/Centcomm airlocks and firedoors now block explosions as walls do
/:cl:

~~@ChangelingRain's idea~~
@optimumtact's idea

Cyberboss committed:
Attempted fix for explosion dodging

optimumtact commented:
10:00 <%oranges> that is the single worst idea I have ever seen
10:00 <%oranges> simply remember the outside line of turfs that exploded
10:00 <%oranges> and apply explosion damage to any mob that passes through them
10:00 <%oranges> duh

Cyberboss commented:
wew

Cyberboss commented:
Actually leaving this open cause it's a good base, I'll finish it tonight

optimumtact commented:
10:03 <%oranges> I believe in you cy
10:03 <%oranges> ber
10:03 <%oranges> boss
10:03 <%oranges> he's not even in here

Cyberboss commented:
I'm at work lol

anconfuzedrock commented:
this keeps slow booms right

Cyberboss commented:
@anconfuzedrock All this is changing is that things that enter previously exploded tiles will get exploded. No more running to the epicenter to dodge it

Cyberboss committed:
Refactor turf/ex_act to return the new tuf

Cyberboss committed:
Fixed a dumb

Cyberboss committed:
Fixes more dumbs

Cyberboss committed:
@optimumtact's suggestion

Cyberboss committed:
If it blows up an rwall, it should blow up a door

Cyberboss committed:
Gives firedoors some explosion_block

Cyberboss committed:
Remove debugging

Cyberboss committed:
Remove more debugging

Cyberboss committed:
Fixed a missing ex_act return

Cyberboss committed:
Merge branch 'Cyberboss-patch-1' of https://github.com/Cyberboss/tgstation

Cyberboss committed:
Let's limit this to mobs

Because that's all we care about

Cyberboss committed:
We need to keep this check

Otherwise the proc may crashe and turfs could end up constantly exploding, gibbing anyone who walks through them

Cyberboss committed:
Buffs diamond, centcomm, and external airlocks

Cyberboss committed:
Bigger block for heavy firedoors

MrStonedOne commented on code/game/objects/explosion.dm:
```diff
@@ -131,13 +132,19 @@
+				var/t_type = T.type
+				var/turf/NT = T.ex_act(dist)
+				if(!istype(NT))
+					WARNING("Turf ex_act for type [t_type] did not return the new turf")
```
what the fuck is all of this about.

MrStonedOne replied:
Turfs are static references.

if I store a reference to a floor turf, and later on, that turf becomes a space, that reference now holds the space turf. This is automatic in byond because with the exception of temp turf (turfs created without a loc), turf references are just their x/y/z and when accessing a turf it resolves to a lookup in the game world's 3dim array of turfs.

MrStonedOne commented on code/game/turfs/turf.dm:
```diff
@@ -111,6 +112,9 @@
 		B.HasProximity(AM)
 
+	if(explosion_level && ismob(AM))
+		AM.ex_act(explosion_level)
```
so when you enter a turf, you go boom, then when you enter it again, before the explosion wakes up, you go boom again, rinse repeat for who knows how long when explosions take a fair amount of time to process.

A mob shouldn't be getting hit by the same explosion twice, your change would make that happen. You are gonna need to find some way to track the mob.

Also, this wouldn't work if two explosions happen at the same time, such as exploding 5 tank transfer valves via signal that are all in the same backpack.

And this wouldn't work with one explosion because fuel tanks. the fuel tank would explode, overriding the vars you set on the turf.

MrsTonedOne reviewed (changes requested):
remove all of these needless changes to ex_act so the diff is much easier to read, figure out a way to make this work when multiple explosions go off at once or an explosion triggers another explosion, and figure out a way to prevent really laggy explosions from hitting a mob multiple times.

ChangelingRain commented:
I have a prediction
the prediction is you end up using my solution

Cyberboss committed:
Revert "We need to keep this check"

This reverts commit 47029c9ba02b0332610ec0bac2c908ce1181a01c.

Cyberboss committed:
Revert "Remove more debugging"

This reverts commit b598fb795b79806fc9b4d38919a1f9fb17d29299.

Cyberboss committed:
Revert "Remove debugging"

This reverts commit 68a6bd1c6e8e218dd5dce248c79bf184b4622e2d.

Cyberboss committed:
Revert "Fixed a missing ex_act return"

This reverts commit c3dbe3b99b9035bac8be2930110f9b6db8acec37.

Cyberboss committed:
Revert "Fixes more dumbs"

This reverts commit dc5767f2f2cdcd410afbd46c9fcecfdf7701f028.

Cyberboss committed:
Revert "Fixed a dumb"

This reverts commit c447cfdd8c92a4af61bf776ea11500f9aa52fc99.

Cyberboss committed:
Revert "Refactor turf/ex_act to return the new tuf"

This reverts commit 38ddfca1800dbc67a59e617bd52924f78a17666e.

Cyberboss committed:
The new old fashioned way

Cyberboss committed:
Compile before you push you slitherbrain

Cyberboss committed:
Merge upstream

Cyberboss committed:
Fixes and solidarity

Cyberboss committed:
Only one circumference please

AnturK commented:
Was this testmerged ? Feels like something that should be.

MrStonedOne commented:
Broke explosions.

People reported being randomly hit by explosions when there was no sign an explosion was going on near them.

Cyberboss commented:
I know what the problem is. I didn't clear the list after the explosion,
only when it lags.

On Jan 2, 2017 8:24 AM, "Kyle Spier-Swenson" <notifications@github.com>
wrote:

> Broke explosions.
>
> People reported being randomly hit by explosions when there was no sign an
> explosion was going on near them.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/22360#issuecomment-269972027>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHywelSiPt03_isGJ-qI1XcTMJ0YDoPvks5rOPpwgaJpZM4LSS2g>
> .
>

Cyberboss committed:
Fixed

Cyberboss commented:
Test moige again pls

AnturK commented:
Conflicts, was this testmerged again ?

Cyberboss committed:
World's shittiest conflict

optimumtact commented:
I've asked a few times but can't get a hold of anyone

AnturK commented:
Was tested without any issues according to MSO so merging this now.

coiax commented:
@Cyberboss your changelog is super vague and non-specific, and people keep asking in OOC what it means.

optimumtact commented:
previously under serious lag you could move through the edge of an explosion in a tick where the tile was not processed, thereby avoiding an explosion.

This pr changes that so that attempting to move through a tile that is part of the edge of a current explosion will apply that damage to you

MrStonedOne commented:
Actually, i think they might be talking about the bottom two entries.

optimumtact commented:
well they seem pretty clear to me

coiax commented:
Talking about the explosion dodging by moving towards the epicentre. I am aware of the bug and the fix, I'm just passing on the chatter from ooc.
